tests/Aspire.Hosting.Tests/Dashboard/ResourcePublisherTests.cs

```diff
@@ -100,19 +123,35 @@ public async Task MergesResourcesInSnapshot()
 
         await publisher.IntegrateAsync(a1, ResourceChangeType.Upsert).ConfigureAwait(false);
         await publisher.IntegrateAsync(a2, ResourceChangeType.Upsert).ConfigureAwait(false);
+
+        var (snapshot, subscription) = publisher.Subscribe();
+
+        Assert.Same(a1, Assert.Single(snapshot));
+
+        var received = false;
+        var task = Task.Run(async () =>
+        {
+            await foreach (var _ in subscription.WithCancellation(cts.Token).ConfigureAwait(false))
+            {
+                received = true;
+            }
+        });
+
         await publisher.IntegrateAsync(a3, ResourceChangeType.Upsert).ConfigureAwait(false);
 
-        var (snapshot, _) = publisher.Subscribe();
+        await Task.Delay(50);
```

Comment:
Yuck.

I've noticed a number of places our async code relies on a cancellation token to abort channels and pipes. It would be nice if we had a mechanism to signal the publisher was complete and subscriptions gracefully complete. e.g. the publisher was told to complete, subscription finishes draining, the await foreach returns false and the task completes.

Benefits:
1. Graceful completion is faster than aborting cancellation tokens. Probably not impactful.
2. This test could await the task being completed and then assert received is false.

Reply:
Can you explain more how signalling completion would help? The lifetime of the connection is owned by the client, not the server, so it's the dashboard that initiated closure, using cancellation. At that point, any buffered messages are dropped. Cancellation should be effectively immediate too, which helps responsiveness.

Reply:
Agree that the delay in the test isn't ideal. I thought about it for a while, but it's a tricky case. After a while I timed out and moved on.

Reply:
That said, xUnit runs tests concurrently and any lengthy wait that isn't touching CPU might have a larger wall time, but shouldn't increase overall time for most runs.

Reply:
I checked out your branch and played around with it. I have some improvements that make testing better and I fixed a memory leak: https://github.com/dotnet/aspire/compare/suppress-sending-unchanged-data

* `ResourcePublisher` is [relying on a CT to stop a subscription](https://github.com/dotnet/aspire/blob/375645de437076c2a1a3b0fc643342d5f0b9d369/src/Aspire.Hosting/Dashboard/ResourcePublisher.cs#L45-L55), but the CT is coming from the constructor and has a lifetime of the Blazor session. The CT passed to the enumerable with `WithCancellation` isn't used. That means as you navigate around the app, the outgoingChannels collection will keep growing. Fix by adding `[EnumeratorCancellation]` parameter to the method.
* ResourcePublisher now implements IDisposable. Disposing the publisher marks all the channels as complete. Tests can dispose the publisher and read to the end of the subscription to verify something hasn't been written.

> The lifetime of the connection is owned by the client, not the server, so it's the dashboard that initiated closure, using cancellation.

A test just needs to read the subscription to the end and not request cancellation. The missing part was a way for the publisher to mark its subscriptions as ending.